src/main/java/org/manifold/compiler/front/ExpressionGraph.java

```diff
@@ -201,8 +209,12 @@ public void addSubExpressionGraph(ExpressionGraph subGraph,
 
     // each edge in subgraph -> edge in main graph should refer to the same source/target
     subGraph.getEdges().forEach(edge -> {
-        exprEdgeMap.get(edge).setSource(exprVertexMap.get(edge.getSource()));
-        exprEdgeMap.get(edge).setTarget(exprVertexMap.get(edge.getTarget()));
+        if (edge.getSource() != null) {
```

Comment:
@maxqchen - I'm not as familiar with function elaboration as you, so you should still look at the rest of this.